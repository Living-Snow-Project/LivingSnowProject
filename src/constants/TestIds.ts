const TestIds = {
  GPS: {
    gpsManualInputTestId: "gps-manual-input",
    gpsManualPressableTestId: "gps-manual-pressable",
  },
  Photos: {
    photoSelectorTestId: "photo-selector",
  },
  Pickers: {
    recordSelectorTestId: "record-type-picker",
    atlastSelectorTestId: "atlas-type-picker",
  },
  RecordScreen: {
    UploadButton: "record-screen-upload-button",
  },
  TimelineScreen: {
    RefreshControl: "timeline-screen-refresh-control",
    SettingsButton: "timeline-screen-settings-button",
    NewRecordButton: "timeline-screen-record-button",
  },
  SettingsScreen: {
    ShowGpsWarning: "settings-screen-show-gps-warning",
    ShowAtlasRecords: "settings-screen-show-atlas-records",
    ShowOnlyAtlasRecords: "settings-screen-show-only-atlas-records",
  },
  Icons: {
    SampleIcon: "sample-icon",
    SightingIcon: "sighting-icon",
    PictureIcon: "picture-icon",
    DeleteIcon: "delete-icon",
    EditIcon: "edit-icon",
    SnowIcon: "snow-icon",
  },
  RecordList: {
    EditRecordAction: "edit-record-action",
    DeleteRecordAction: "delete-record-action",
  },
};

export default TestIds;
