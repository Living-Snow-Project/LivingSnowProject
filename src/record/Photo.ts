const randomInteger = (): number => Math.floor(Math.random() * 1_000_000);

const makeExamplePhoto = ({
  isLocal = false,
  uri = `${randomInteger()}`,
  size = randomInteger(),
  width = randomInteger(),
  height = randomInteger(),
} = {}): Photo => ({
  uri: isLocal ? `file://${uri}` : uri,
  size,
  width,
  height,
});

const makeExamplePendingPhoto = ({
  isLocal = false,
  id = randomInteger(),
  uri = `${randomInteger()}`,
  size = randomInteger(),
  width = randomInteger(),
  height = randomInteger(),
} = {}): PendingPhoto => ({
  id,
  ...makeExamplePhoto({ isLocal, uri, size, width, height }),
});

export { makeExamplePhoto, makeExamplePendingPhoto };
