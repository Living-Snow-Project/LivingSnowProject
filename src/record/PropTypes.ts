import PropTypes from "prop-types";

const AlgaeRecordTypePropType: PropTypes.Requireable<AlgaeRecordType> =
  PropTypes.oneOf<AlgaeRecordType>([
    "Sample",
    "Sighting",
    "Atlas: Red Dot",
    "Atlas: Red Dot with Sample",
    "Atlas: Blue Dot",
    "Atlas: Blue Dot with Sample",
    "Undefined",
  ]);

const AtlasTypePropType: PropTypes.Requireable<AtlasType> =
  PropTypes.oneOf<AtlasType>([
    "Ash",
    "Dirt or Debris",
    "Forest or Vegetation",
    "Mix of Algae and Dirt",
    "Other",
    "Snow Algae",
    "White Snow",
    "Undefined",
  ]);

// TODO: const AlgaeRecordPropType: PropTypes.Requireable<AlgaeRecord>...
// backend types will need to change first (ie. lat, long are strings in db model)
const AlgaeRecordPropType = PropTypes.shape({
  // TODO: pendingRecords uses uuidv4() to generate an id...
  id: PropTypes.oneOfType([
    PropTypes.number.isRequired,
    PropTypes.string.isRequired,
  ]).isRequired,
  type: AlgaeRecordTypePropType.isRequired,
  name: PropTypes.string,
  organization: PropTypes.string,
  date: PropTypes.instanceOf(Date).isRequired,
  latitude: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  longitude: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  tubeId: PropTypes.string,
  locationDescription: PropTypes.string,
  notes: PropTypes.string,
  photoUris: PropTypes.string,
  atlasType: AtlasTypePropType,
});

export { AlgaeRecordPropType, AlgaeRecordTypePropType, AtlasTypePropType };
