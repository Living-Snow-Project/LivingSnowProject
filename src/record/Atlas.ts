import { AtlasDescription } from "../constants/Strings";

enum AtlasType {
  Undefined = -1,
  SnowAlgae,
  DirtOrDebris,
  Ash,
  WhiteSnow,
  MixOfAlgaeAndDirt,
  ForestOrVegetation,
  Other,
  Max,
}

type AtlasPickerItem = {
  value: AtlasType;
  label: string;
};

// specific format for RNPickerSelect
const AtlasPickerItems: AtlasPickerItem[] = [
  { value: AtlasType.SnowAlgae, label: AtlasDescription.SnowAlgae },
  { value: AtlasType.DirtOrDebris, label: AtlasDescription.DirtOrDebris },
  { value: AtlasType.Ash, label: AtlasDescription.Ash },
  { value: AtlasType.WhiteSnow, label: AtlasDescription.WhiteSnow },
  {
    value: AtlasType.MixOfAlgaeAndDirt,
    label: AtlasDescription.MixOfAlgaeAndDirt,
  },
  {
    value: AtlasType.ForestOrVegetation,
    label: AtlasDescription.ForestOrVegetation,
  },
  { value: AtlasType.Other, label: AtlasDescription.Other },
];

const getAtlasPickerItem = (type: AtlasType): AtlasPickerItem =>
  type > AtlasType.Undefined && type < AtlasType.Max
    ? AtlasPickerItems[type]
    : { value: AtlasType.Undefined, label: AtlasDescription.Undefined };

const getAllAtlasPickerItems = (): AtlasPickerItem[] => AtlasPickerItems;

export { AtlasType, getAtlasPickerItem, getAllAtlasPickerItems };
