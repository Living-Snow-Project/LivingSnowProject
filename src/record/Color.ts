import { AlgaeColorDescription } from "../constants/Strings";

type AlgaeColorPickerItem = {
  value: AlgaeColor;
  label: string;
};

// specific format for RNPickerSelect
const algaeColorPickerItems: AlgaeColorPickerItem[] = [
  { value: "Select a color", label: AlgaeColorDescription.Select },
  { value: "Red", label: AlgaeColorDescription.Red },
  { value: "Pink", label: AlgaeColorDescription.Pink },
  { value: "Grey", label: AlgaeColorDescription.Grey },
  { value: "Green", label: AlgaeColorDescription.Green },
  { value: "Orange", label: AlgaeColorDescription.Orange },
  { value: "Yellow", label: AlgaeColorDescription.Yellow },
  { value: "Other", label: AlgaeColorDescription.Other },
];

const getAlgaeColorPickerItem = (color: AlgaeColor): AlgaeColorPickerItem => {
  const result = algaeColorPickerItems.find((cur) => cur.value === color);

  if (result === undefined) {
    return { value: "Other", label: AlgaeColorDescription.Other };
  }

  return result;
};

const getAllAlgaeColorPickerItems = (): AlgaeColorPickerItem[] =>
  algaeColorPickerItems;

export { getAlgaeColorPickerItem, getAllAlgaeColorPickerItems };
