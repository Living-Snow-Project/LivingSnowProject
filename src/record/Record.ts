import { RecordDescription } from "../constants/Strings";

type RecordTypePickerItem = {
  value: AlgaeRecordType;
  label: string;
};

// specific format for RNPickerSelect
const recordTypePickerItems: RecordTypePickerItem[] = [
  { value: "Sample", label: RecordDescription.Sample },
  { value: "Sighting", label: RecordDescription.Sighting },
  { value: "Atlas: Red Dot", label: RecordDescription.AtlasRedDot },
  {
    value: "Atlas: Red Dot with Sample",
    label: RecordDescription.AtlasRedDotWithSample,
  },
  { value: "Atlas: Blue Dot", label: RecordDescription.AtlasBlueDot },
  {
    value: "Atlas: Blue Dot with Sample",
    label: RecordDescription.AtlasBlueDotWithSample,
  },
];

const getAllRecordTypePickerItems = (): RecordTypePickerItem[] =>
  recordTypePickerItems;

const getRecordTypePickerItem = (
  type: AlgaeRecordType
): RecordTypePickerItem => {
  const result = recordTypePickerItems.find((cur) => cur.value === type);

  if (result === undefined) {
    return { value: "Undefined", label: RecordDescription.Undefined };
  }

  return result;
};

const isSample = (type: AlgaeRecordType): boolean =>
  Array<AlgaeRecordType>(
    "Sample",
    "Atlas: Red Dot with Sample",
    "Atlas: Blue Dot with Sample"
  ).includes(type);

const isAtlas = (type: AlgaeRecordType): boolean =>
  Array<AlgaeRecordType>(
    "Atlas: Red Dot",
    "Atlas: Red Dot with Sample",
    "Atlas: Blue Dot",
    "Atlas: Blue Dot with Sample"
  ).includes(type);

// TODO: makeExampleRecord should also be used to seed RecordScreen (ie. no more "no records to display")
const makeExampleRecord = (type: AlgaeRecordType): AlgaeRecord => {
  const atlasType: AtlasType = isAtlas(type) ? "Snow Algae" : "Undefined";

  return {
    id: 1234,
    type,
    name: "test name",
    date: new Date("2021-09-16T00:00:00"),
    organization: "test org",
    latitude: -123.456,
    longitude: 96.96,
    tubeId: isSample(type) ? "LAB-1337" : "",
    locationDescription: "test location",
    notes: "test notes",
    atlasType,
    photos: [
      {
        uri: "46",
        size: 100,
        width: 16,
        height: 16,
      },
      { uri: "23", size: 4096, width: 128, height: 128 },
    ],
  };
};

// needed for JSON.parse, otherwise date property will become string
const recordReviver = (key: string, value: any): any => {
  if (key === "date") {
    return new Date(value);
  }

  return value;
};

// decodes AlgaeRecord or AlgaeRecord[] JSON
function jsonToRecord<T>(json: string): T {
  return JSON.parse(json, recordReviver);
}

// want to display date in YYYY-MM-DD format
const recordDateFormat = (date: Date): string => {
  const dayNum: number = date.getDate();
  let day: string = `${dayNum}`;

  if (dayNum < 10) {
    day = `0${dayNum}`;
  }

  const monthNum: number = date.getMonth() + 1;
  let month: string = `${monthNum}`;

  if (monthNum < 10) {
    month = `0${monthNum}`;
  }

  return `${date.getFullYear()}-${month}-${day}`;
};

export {
  getRecordTypePickerItem,
  getAllRecordTypePickerItems,
  makeExampleRecord,
  isSample,
  isAtlas,
  jsonToRecord,
  recordDateFormat,
};
