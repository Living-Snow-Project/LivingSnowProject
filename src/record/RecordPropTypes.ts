import PropTypes from "prop-types";
import { AtlasType } from "./Atlas";

const AlgaeRecordTypePropType: PropTypes.Requireable<AlgaeRecordType> =
  PropTypes.oneOf<AlgaeRecordType>([
    "Sample",
    "Sighting",
    "Atlas: Red Dot",
    "Atlas: Red Dot with Sample",
    "Atlas: Blue Dot",
    "Atlas: Blue Dot with Sample",
    "Undefined",
  ]);

// TODO: const AlgaeRecordPropType: PropTypes.Requireable<AlgaeRecord>...
// backend types will need to change first (ie. lat, long are strings in db model)
const AlgaeRecordPropType = PropTypes.shape({
  id: PropTypes.number.isRequired,
  type: AlgaeRecordTypePropType.isRequired,
  name: PropTypes.string,
  organization: PropTypes.string,
  date: PropTypes.instanceOf(Date).isRequired,
  latitude: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  longitude: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  tubeId: PropTypes.string,
  locationDescription: PropTypes.string,
  notes: PropTypes.string,
  photoUris: PropTypes.string,
  atlasType: PropTypes.oneOf(Object.values(AtlasType)).isRequired,
});

export { AlgaeRecordPropType, AlgaeRecordTypePropType };
