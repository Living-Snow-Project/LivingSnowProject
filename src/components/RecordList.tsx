import React, { useCallback } from "react";
import { Text, View } from "react-native";
import PropTypes from "prop-types";
import { AlgaeRecordPropType } from "../record/RecordPropTypes";
import styles from "../styles/Timeline";
import TimelineRow from "./TimelineRow";

type RecordListProps = {
  records: AlgaeRecord[];
  header: string;
  omitRecord?: (record: AlgaeRecord) => boolean;
};

export default function RecordList({
  records,
  header,
  omitRecord,
}: RecordListProps) {
  const RenderRecords = useCallback(
    () => (
      <>
        {records.map((record) => {
          // @ts-ignore (TS complains it is possibly undefined however propTypes.defaultProps assigns it a value)
          if (omitRecord(record)) {
            return null;
          }
          return <TimelineRow key={record.id} record={record} />;
        })}
      </>
    ),
    [records, omitRecord]
  );

  if (records.length === 0) {
    return null;
  }

  return (
    <>
      <View style={styles.recordStatusContainer}>
        <Text style={styles.recordStatusText}>{header}</Text>
      </View>
      <RenderRecords />
    </>
  );
}

RecordList.propTypes = {
  records: PropTypes.arrayOf(AlgaeRecordPropType).isRequired,
  header: PropTypes.string.isRequired,
  omitRecord: PropTypes.func,
};

RecordList.defaultProps = {
  omitRecord: () => false,
};
