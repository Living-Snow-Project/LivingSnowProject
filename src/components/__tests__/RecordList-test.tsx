import React from "react";
import { Alert } from "react-native";
import { fireEvent, render, waitFor } from "@testing-library/react-native";
import { PendingRecordList } from "../RecordList";
import { makeExampleRecord } from "../../record/Record";
import TestIds from "../../constants/TestIds";

describe("PendingRecordList tests", () => {
  test("renders", () => {
    const { toJSON } = render(
      <PendingRecordList
        records={[makeExampleRecord("Sighting")]}
        onUpdate={() => {}}
      />
    );

    expect(toJSON()).toMatchSnapshot();
  });

  test("confirm delete record", async () => {
    let okPressed = false;
    const onUpdateMock = jest.fn();
    const alertMock = jest
      .spyOn(Alert, "alert")
      .mockImplementationOnce((title, message, buttons) => {
        // @ts-ignore
        buttons[0].onPress(); // "Yes" button
        okPressed = true;
      });

    const { getByTestId } = render(
      <PendingRecordList
        records={[makeExampleRecord("Sighting")]}
        onUpdate={onUpdateMock}
      />
    );

    fireEvent.press(getByTestId(TestIds.RecordList.DeleteRecordAction));
    await waitFor(() => expect(okPressed).toBe(true));

    expect(alertMock).toBeCalledTimes(1);
    expect(onUpdateMock).toBeCalled();
    alertMock.mockReset();
  });

  test("cancel delete record", () => {
    const noButtonMock = jest.fn();
    const alertMock = jest
      .spyOn(Alert, "alert")
      .mockImplementationOnce((title, message, buttons) => {
        // @ts-ignore
        noButtonMock({ text: buttons[1].text, style: buttons[1].style }); // "No" button
      });

    const { getByTestId } = render(
      <PendingRecordList
        records={[makeExampleRecord("Sighting")]}
        onUpdate={() => {}}
      />
    );

    fireEvent.press(getByTestId(TestIds.RecordList.DeleteRecordAction));
    expect(alertMock).toBeCalledTimes(1);
    expect(noButtonMock).toHaveBeenLastCalledWith({
      text: "No",
      style: "cancel",
    });
  });

  test("edit record", () => {
    const { getByTestId } = render(
      <PendingRecordList
        records={[makeExampleRecord("Sighting")]}
        onUpdate={() => {}}
      />
    );

    fireEvent.press(getByTestId(TestIds.RecordList.EditRecordAction));
  });
});
