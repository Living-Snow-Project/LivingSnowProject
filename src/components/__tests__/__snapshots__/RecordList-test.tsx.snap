// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PendingRecordList tests renders 1`] = `
Array [
  <View
    style={
      Object {
        "backgroundColor": "lightgrey",
        "borderBottomWidth": 1,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 16,
          "textAlign": "center",
        }
      }
    >
      Pending
    </Text>
  </View>,
  <View
    collapsable={false}
    forwardedRef={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onLayout={[Function]}
    style={
      Object {
        "overflow": "hidden",
      }
    }
  >
    <View
      style={
        Object {
          "bottom": 0,
          "flexDirection": "row-reverse",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": Array [
            Object {
              "translateX": -10000,
            },
          ],
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "width": 128,
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Array [
                  Object {
                    "alignItems": "center",
                    "borderBottomWidth": 1,
                    "borderLeftWidth": 1,
                    "height": "100%",
                    "justifyContent": "center",
                  },
                  Object {
                    "backgroundColor": "lightgrey",
                  },
                ],
                Object {
                  "opacity": 1,
                },
              ]
            }
            testID="delete-record-action"
          >
            <Text />
          </View>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Array [
                  Object {
                    "alignItems": "center",
                    "borderBottomWidth": 1,
                    "borderLeftWidth": 1,
                    "height": "100%",
                    "justifyContent": "center",
                  },
                  Object {
                    "backgroundColor": "gainsboro",
                  },
                ],
                Object {
                  "opacity": 1,
                },
              ]
            }
            testID="edit-record-action"
          >
            <Text />
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
      />
    </View>
    <View
      collapsable={false}
      forwardedRef={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      pointerEvents="auto"
      style={
        Object {
          "transform": Array [
            Object {
              "translateX": -0,
            },
          ],
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "white",
            "borderBottomWidth": 1,
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          testID="1234"
        >
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "row",
              }
            }
          >
            <View
              style={
                Object {
                  "flex": 0.7,
                  "marginLeft": 1,
                }
              }
            >
              <Text>
                2021-09-16
test name (test org)
Coordinates: -123.456, 96.96
              </Text>
            </View>
            <View
              style={
                Object {
                  "flex": 0.15,
                  "marginLeft": 1,
                }
              }
            >
              <Text />
            </View>
            <View
              style={
                Object {
                  "flex": 0.15,
                  "marginLeft": 1,
                }
              }
            >
              <Text />
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "alignSelf": "flex-end",
                    "backgroundColor": "red",
                    "borderRadius": 40,
                    "position": "absolute",
                    "right": 2,
                    "top": 1,
                    "width": "33%",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "white",
                    }
                  }
                >
                  2
                </Text>
              </View>
            </View>
          </View>
          <Text>
            Location Description: test location
Notes: test notes
          </Text>
        </View>
      </View>
    </View>
  </View>,
]
`;
